feat: Add support for logging to file using -lf option and moved arguments parser into individual function

Am0stafa opened the pull request:
## Description

This pull request adds a new feature to the CLI tool that allows users to redirect all logs to a file using the `-lf` flag. When the `-lf` flag is specified, all logs will be written to a file named "logs.log" in the root directory of the tool.

If the "logs.log" file already exists, new logs will be appended to the end of the file. If the file doesn't exist, it will be created automatically.

This change is helpful for several reasons:

- It gives users more control over how they want to manage and analyze the logs generated by the tool.
- It makes it easier to troubleshoot issues that arise during runtime since all the logs are stored in a single file that can be easily accessed and analyzed.
- It allows users to run the tool in the background without worrying about console output.
- It can help reduce the amount of clutter in the console output, especially when running the tool for long periods of time.
- It enables the tool to be used in environments where console output is not desirable or possible, such as when running the tool in a Docker container.
<br>

This pull request also moves the argument parser for the cli tool from `scanner.py` into a new module named `argument.py`. The argument parser has also been refactored into a separate function named `arg_parser()`.

By separating the argument parser into its own module and function, the code is now more modular and easier to maintain. This change also enables easy modification and extension of the command-line argument parser, making it easier to add or modify command-line arguments in the future.

The `arg_parser()` function takes no arguments and returns an instance of `argparse.Namespace` containing the parsed arguments. It has been designed to be simple and easy to use, with clear documentation and helpful error messages.


## Type of Change

- [x] Feature addition
- [x] Refactoring

## Checklist

- [x] Added support for logging to file using `-lf` flag.
- [x] Created logs file in the user-specified file/directory even if it doesn't exist.
- [x] Appended new logs to the end of the file if it already exists.
- [x] Updated documentation to reflect the new feature.
- [x] Tested the feature on multiple platforms.
- [x] Moved the argument parser into a separate module.
- [x] Created a new module named `argument.py` to handle argument parsing.
- [x] Refactored the argument parsing logic into a separate function named `parser()`.
- [x] Documented the `parser()` function and its usage.
- [x] Tested the feature

## Overall

This change improves the maintainability and flexibility of the codebase by separating the argument parsing logic into its own module and function. It also makes the code more modular and easier to understand, which can help with future development and debugging.

Am0stafa commented:
## Description

This pull request also moves the argument parser for the cli tool from `scanner.py` into a new module named `argument.py`. The argument parser has also been refactored into a separate function named `parser()`.

By separating the argument parser into its own module and function, the code is now more modular and easier to maintain. This change also enables easy modification and extension of the command-line argument parser, making it easier to add or modify command-line arguments in the future.

The `parser()` function takes no arguments and returns an instance of `argparse.Namespace` containing the parsed arguments. It has been designed to be simple and easy to use, with clear documentation and helpful error messages.

## Type of Change

- [x] Refactoring

## Checklist

- [x] Moved the argument parser into a separate module.
- [x] Created a new module named `argument.py` to handle argument parsing.
- [x] Refactored the argument parsing logic into a separate function named `parser()`.
- [x] Updated the `scanner.py` module to import and use the `parser()` function.
- [x] Documented the `parser()` function and its usage.
- [x] Tested the feature on multiple platforms.

## Overall

This change improves the maintainability and flexibility of the codebase by separating the argument parsing logic into its own module and function. It also makes the code more modular and easier to understand, which can help with future development and debugging.

mshudrak commented on src/gcp_scanner/arguments.py:
```diff
@@ -0,0 +1,122 @@
+import argparse
+import logging
+
+def arg_parser():
```
Provide function description similar to what we have for other functions.

mshudrak commented on src/gcp_scanner/scanner.py:
```diff
@@ -299,105 +299,21 @@ def main():
-                      datefmt='%Y-%m-%d %H:%M:%S')
-
+  if args.log_file:
+    logging.basicConfig(filename='../logs.log',
```
I'd suggest to let user specify logdir path instead of making assumption that parent dir is available for writing.

Am0stafa replied:
I am working on it such that the user specifies the logging directory and if it doesn't exist I create it and put the logs in it.
-ld logsfile
this will create the logsfile directory and adds the logs to it

mshudrak commented on src/gcp_scanner/scanner.py:
```diff
@@ -299,105 +299,21 @@ def main():
+                        format='%(asctime)s - %(levelname)s - %(message)s',
+                        datefmt='%Y-%m-%d %H:%M:%S')
+  else:
+    logging.basicConfig(level=getattr(logging, args.log_level.upper(), None),
```
You can potentially avoid entire if/else statement if you specify filename=None to have logging into the console. I am not 100% sure but worth checking.

Am0stafa replied:
I am working on it such that the user specifies the output file and if it doesn't exist i create it and put the logs in it.
-lf logsfile
this will create the logs file and adds the logs to it

Am0stafa replied:
yeah that actually work

mshudrak reviewed (changes requested):
Thanks for the commit. PTAL at my suggestions.

Am0stafa commented:
@mshudrak all requested changes are done, thanks for the guidance.

mshudrak commented on src/gcp_scanner/arguments.py:
```diff
@@ -0,0 +1,132 @@
+      '--log-dir',
+      default=None,
+      dest='log_directory',
+      help='Save logs to the directory specified rather than displaying in\
```
Let's just have a file that user specify with full/relative path instead of directory.

Am0stafa replied:
a small clarification, the user will specify the full path such as:
-ld logFile/logs.log
-ld logs.log
and so one?

I made it by making the user just specify the logging directory just like the key and output for example so just to be clear do you want it just like the above examples?

Am0stafa replied:
yeah that would be a better approach

mshudrak commented on src/gcp_scanner/scanner.py:
```diff
@@ -299,104 +299,26 @@ def main():
+  if args.log_directory and args.log_directory[0] == '/':
+    args.log_directory = args.log_directory[1:]
+
+  outputDir = f'{args.log_directory}' if args.log_directory else None
```
argparser will assign `log_directory = None` by default You don't need at all.

mshudrak commented on src/gcp_scanner/scanner.py:
```diff
@@ -299,104 +299,26 @@ def main():
+    args.log_directory = args.log_directory[1:]
+
+  outputDir = f'{args.log_directory}' if args.log_directory else None
+  outputFile = f'{args.log_directory}/logs.log' if args.log_directory else None
```
See above. Let's have a log file instead of a directory.

mshudrak commented on src/gcp_scanner/scanner.py:
```diff
@@ -299,104 +299,26 @@ def main():
+  outputDir = f'{args.log_directory}' if args.log_directory else None
+  outputFile = f'{args.log_directory}/logs.log' if args.log_directory else None
+
+  if outputDir:
```
We can capture that in the argument description that it will create a directory/directories for the path provided.

mshudrak commented on src/gcp_scanner/scanner.py:
```diff
@@ -299,104 +299,26 @@ def main():
 
-  logging.basicConfig(level=getattr(logging, args.log_level.upper(), None),
+  # logging to a directory functionality
+  if args.log_directory and args.log_directory[0] == '/':
```
Why do you need that?

Am0stafa replied:
when I was testing what I notest was if the user specify -ld /logFile that would throw an error

mshudrak reviewed (changes requested):
Left some more comments. PTAL.

Am0stafa commented:
@mshudrak I made all the changes requested along with adding support for if the user-specified file or directory doesn't exist we create it, also I cleaned the code a bit and updated the docs and finally I left the removal of "/" to handle errors
now you can:

-lf logs.log
-lf logFile/logs.log
-lf logsFile/test/logs.log

and soo on after that if the file exit it will append to it else it will create it and wright the logs to it

ZetaTwo commented:
I don't think the tool should be creating the directories. This will probably cause more problems than it solves. It is better if it tries to use the path at startup and then error if it doesn't work.

Am0stafa commented:
@ZetaTwo @mshudrak, Thank you for taking the time to review my pull request and provide feedback. I have made the requested change regarding directory creation and also addressed the other changes you suggested. Your feedback was valuable in helping me improve the quality of my code. I appreciate your input and thank you again for your time!

mshudrak commented on src/gcp_scanner/scanner.py:
```diff
@@ -299,104 +299,18 @@ def main():
     force_projects_list = args.force_projects.split(',')
 
-  logging.basicConfig(level=getattr(logging, args.log_level.upper(), None),
+  # logging to a directory functionality
```
I think we can remove 308 and 309.

mshudrak reviewed (changes requested):
Almost ready, PTAL at my comment.

mshudrak commented on src/gcp_scanner/scanner.py:
```diff
@@ -299,104 +299,16 @@ def main():
     force_projects_list = args.force_projects.split(',')
 
-  logging.basicConfig(level=getattr(logging, args.log_level.upper(), None),
+  logging.basicConfig(level=logging.INFO,
```
I don't understand this change. Could you explain why do you want to always have .INFO for basicConfig?

Am0stafa replied:
sorry forgot to return it back😅

mshudrak commented on src/gcp_scanner/scanner.py:
```diff
@@ -299,104 +299,16 @@ def main():
+  logging.basicConfig(level=logging.INFO,
                       format='%(asctime)s - %(levelname)s - %(message)s',
-                      datefmt='%Y-%m-%d %H:%M:%S')
+                      datefmt='%d-%b-%y %H:%M:%S',
```
Let's keep old datefmt.
